aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -98,6 +101,80 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
   return at::arange(like.size(dim), like.options().dtype(at::kLong));
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~ complex / complex_polar ~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+TensorIterator complex_op(
+    Tensor& out,
+    const Tensor& a,
+    const Tensor& b,
+    bool check_mem_overlap) {
+  return TensorIteratorConfig()
+      .set_check_mem_overlap(check_mem_overlap)
+      .add_output(out)
+      .add_input(a)
+      .add_input(b)
+      .allow_cpu_scalars(true)
+      .check_all_same_dtype(false)
+      .build();
+}
+
+Tensor& complex_out(Tensor& result, const Tensor& real, const Tensor& imag) {
+  TORCH_CHECK(
+      result.scalar_type() == toComplexType(real.scalar_type()),
+      "Expected object of scalar type ",
+      toComplexType(real.scalar_type()),
+      " but got scalar type ",
+      result.scalar_type(),
+      " for argument 'out'");
+  TORCH_CHECK(
+      real.scalar_type() == imag.scalar_type(),
+      "Expected object of scalar type ",
+      real.scalar_type(),
+      " but got scalar type ",
+      imag.scalar_type(),
+      " for argument 'imag'");
+  auto iter = complex_op(result, real, imag, /*check_mem_overlap=*/true);
+  complex_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor complex(const Tensor& real, const Tensor& imag) {
+  c10::TensorOptions options = real.options();
```

Comment:
Per your comments, if you'd like a nicer error message you can TORCH_CHECK the dtypes of the inputs here and in `complex_out` above manually. In particular, you could:

- verify the dtypes of real and imag are the same
- verify the dtype of real is torch.float32 or torch.float64, and if not say that complex doesn't support inputs the a dtype of X (whatever real's dtype is)

I think that would look nicer?

Reply:
Yes but my concern is that the second bullet point "seems" more important. If real is torch.int32 and imag is torch.int64, I think it's better to say "Inputs are not floating type" as opposed to  "Expected object of scalar type torch.int32 but got scalar type torch.int64 for argument imag".